### PR TITLE
📦 NEW: `init` hook to enqueue block scripts

### PR DESCRIPTION
**Pull request by asharirfan**
⚠️ **IMPORTANT:** This is a duplicate pull request for ahmadawais/clai-old#47 

Hey 🙌

After reading the discussing in issue ahmadawais/create-guten-block#37, I have done the required changes to the code in `cgb-scripts/template/src/init.php` ✅ 

→ The scripts are being enqueued using `init` WordPress hook.

alt text

Cheers 🤘

### CODE REVIEW

File: packages/cgb-scripts/template/src/init.php [hunk 1]
```diff
@@ -28,38 +30,36 @@ function <% blockNamePHPLower %>_cgb_block_assets() {
...
- * `wp-blocks`: includes block type registration and related functions.
- * `wp-element`: includes the WordPress Element abstraction for describing the structure of your blocks.
- * `wp-i18n`: To internationalize the block's text.
- *
- * @since 1.0.0
- */
-function <% blockNamePHPLower %>_cgb_editor_assets() {
-	// Scripts.
+	// Block scripts.
 	wp_enqueue_script(
```
Comment (ahmadawais) [inline packages/cgb-scripts/template/src/init.php]: Did you test this? Coz. it's wrong. You're enqueuing the files and then register them. Which is wrong. I hope I get time to write automated tests for this but till then do your own due diligence.

File: packages/cgb-scripts/template/src/init.php [hunk 2]
```diff
@@ -28,38 +30,36 @@ function <% blockNamePHPLower %>_cgb_block_assets() {
...
 		plugins_url( '/dist/blocks.build.js', dirname( __FILE__ ) ), // Block.build.js: We register the block here. Built with Webpack.
-		array( 'wp-blocks', 'wp-i18n', 'wp-element' ), // Dependencies, defined above.
+		array( 'wp-blocks', 'wp-i18n', 'wp-element' ) // Dependencies, defined above.
 		// filemtime( plugin_dir_path( __DIR__ ) . 'dist/blocks.build.js' ), // Version: filemtime — Gets file modification time.
 		true // Enqueue the script in the footer.
 	);
 
-	// Styles.
+	// Block backend styles.
 	wp_enqueue_style(
```
Comment (ahmadawais) [inline packages/cgb-scripts/template/src/init.php]: Same wrong enqueue here.

**Review by ahmadawais — CHANGES_REQUESTED** [review]

File: packages/cgb-scripts/template/src/init.php (hunk 1 above)
Comment (asharirfan) [reply packages/cgb-scripts/template/src/init.php]: You're right about the enqueuing. This is a mistake from my side. Although when I ran the tests, I was registering the scripts first and everything was working fine. Going to correct the PR.

Here is the file of the block that I tested:

```php
<?php
/**
 * Blocks Initializer
 *
 * Enqueue CSS/JS of all the blocks.
 *
 * @since   1.0.0
 * @package CGB
 */

// Exit if accessed directly.
if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

/**
 * Enqueue Gutenberg block assets for both frontend + backend.
 *
 * `wp-blocks`: includes block type registration and related functions.
 * `wp-element`: includes the WordPress Element abstraction for describing the structure of your blocks.
 * `wp-i18n`: To internationalize the block's text.
 *
 * @since 1.0.0
 */
function my_block_cgb_block_assets() {
	// Styles.
	wp_register_style(
		'my_block-cgb-style-css', // Handle.
		plugins_url( 'dist/blocks.style.build.css', dirname( __FILE__ ) ), // Block style CSS.
		array( 'wp-blocks' ) // Dependency to include the CSS after it.
		// filemtime( plugin_dir_path( __DIR__ ) . 'dist/blocks.style.build.css' ) // Version: filemtime — Gets file modification time.
	);

	// Block scripts.
	wp_register_script(
		'my_block-cgb-block-js', // Handle.
		plugins_url( '/dist/blocks.build.js', dirname( __FILE__ ) ), // Block.build.js: We register the block here. Built with Webpack.
		array( 'wp-blocks', 'wp-i18n', 'wp-element' ), // Dependencies, defined above.
		false, // filemtime( plugin_dir_path( __DIR__ ) . 'dist/blocks.build.js' ), // Version: filemtime — Gets file modification time.
		true // Enqueue the script in the footer.
	);

	// Block backend styles.
	wp_register_style(
		'my_block-cgb-block-editor-css', // Handle.
		plugins_url( 'dist/blocks.editor.build.css', dirname( __FILE__ ) ), // Block editor CSS.
		array( 'wp-edit-blocks' ) // Dependency to include the CSS after it.
		// filemtime( plugin_dir_path( __DIR__ ) . 'dist/blocks.editor.build.css' ) // Version: filemtime — Gets file modification time.
	);

	/**
	 * Register the block scripts and styles for both frontend and
	 * backend to ensure they are enqueued when the editor loads.
	 *
	 * @link https://wordpress.org/gutenberg/handbook/blocks/writing-your-first-block-type/
	 */
	register_block_type( 'cgb/block-my-block', array(
		'style' => 'my_block-cgb-style-css', // Enqueue blocks.style.build.css on both frontend & backend.
		'editor_script' => 'my_block-cgb-block-js', // Enqueue blocks.build.js in the editor only.
		'editor_style' => 'my_block-cgb-block-editor-css', // Enqueue blocks.editor.build.css in the editor only.
	) );
} // End function my_block_cgb_block_assets().

// Hook: Block assets.
add_action( 'init', 'my_block_cgb_block_assets' );

```

File: packages/cgb-scripts/template/src/init.php [hunk 3]
```diff
@@ -24,24 +24,24 @@
...
 
 	// Block scripts.
-	wp_enqueue_script(
+	wp_register_script(
 		'<% blockNamePHPLower %>-cgb-block-js', // Handle.
 		plugins_url( '/dist/blocks.build.js', dirname( __FILE__ ) ), // Block.build.js: We register the block here. Built with Webpack.
-		array( 'wp-blocks', 'wp-i18n', 'wp-element' ) // Dependencies, defined above.
-		// filemtime( plugin_dir_path( __DIR__ ) . 'dist/blocks.build.js' ), // Version: filemtime — Gets file modification time.
+		array( 'wp-blocks', 'wp-i18n', 'wp-element' ), // Dependencies, defined above.
+		false, // filemtime( plugin_dir_path( __DIR__ ) . 'dist/blocks.build.js' ), // Version: filemtime — Gets file modification time.
```
Comment (asharirfan) [inline packages/cgb-scripts/template/src/init.php]: Introduced `false` as the 4th parameter because the next parameter – `true // Enqueue the script in the footer.` – was useless because it was being taken as the 4th due to the comment.

Comment (ahmadawais) [reply packages/cgb-scripts/template/src/init.php]: That is not useless. Read this https://github.com/ahmadawais/create-guten-block/pull/43

**Comment by ahmadawais** [comment]
@asharirfan What is the current status of this PR? Did you test it?

**Comment by asharirfan** [comment]
Yes, I have tested it. It is working on my end.

File: packages/cgb-scripts/template/src/init.php (hunk 3 above)
Comment (asharirfan) [reply packages/cgb-scripts/template/src/init.php]: The way I see it, 4th parameter never made it to the master branch. See this merge commit of the PR: https://github.com/ahmadawais/create-guten-block/commit/d506e27b306206ca0b95c434aeb6c3c44c9c86a5#diff-4317d120593768bbdedb26f6949dadad

Related file on master: https://github.com/ahmadawais/create-guten-block/blob/master/packages/cgb-scripts/template/src/init.php#L51-L52

**Comment by asharirfan** [comment]
@ahmadawais Any updates on this one? Last time I tested this PR, it worked.

**Comment by drzraf** [comment]
By default, block's Javascript that takes care of rendering the content is only loaded for editors.
IHMO this is [a bit confusing](https://wordpress.org/support/topic/gutenberg-react-block-frontend-rendering/) for Gutenberg newcomers because it seems to ignore frontend rendering.

**Comment by drdogbot7** [comment]
What if I'm using CGB to create a plugin with multiple block types? The build script only creates one block.build.js. The current init.php just enqueues those build files, but this method seems to imply that styles and JS only apply to specific blocks. Would we register _each_ block type with the same `style`, `editor-script`, and `editor-style`?

**Comment by asharirfan** [comment]
@ahmadawais Tagging you here for your review on this PR.

I have synced this PR with the master branch to avoid merge conflicts.

I have also tested the script by creating a new block. Here are the screenshots:

Editor View
======
image

Frontend View
======
image

Looking forward to your review 🙌

File: packages/cgb-scripts/template/src/init.php (hunk 1 above)
Comment (asharirfan) [reply packages/cgb-scripts/template/src/init.php]: Corrected this mistake.

File: packages/cgb-scripts/template/src/init.php (hunk 2 above)
Comment (asharirfan) [reply packages/cgb-scripts/template/src/init.php]: Corrected this one too.

File: packages/cgb-scripts/template/src/init.php [hunk 4]
```diff
@@ -16,49 +16,50 @@
...
-function <% blockNamePHPLower %>_cgb_editor_assets() { // phpcs:ignore
-	// Scripts.
-	wp_enqueue_script(
+	// Block scripts.
+	wp_register_script(
 		'<% blockNamePHPLower %>-cgb-block-js', // Handle.
 		plugins_url( '/dist/blocks.build.js', dirname( __FILE__ ) ), // Block.build.js: We register the block here. Built with Webpack.
-		array( 'wp-blocks', 'wp-i18n', 'wp-element', 'wp-editor' ), // Dependencies, defined above.
-		// filemtime( plugin_dir_path( __DIR__ ) . 'dist/blocks.build.js' ), // Version: File modification time.
+		array( 'wp-blocks', 'wp-i18n', 'wp-element' ), // Dependencies, defined above.
```
Comment (ahmadawais) [inline packages/cgb-scripts/template/src/init.php]: Why did you remove `'wp-editor'`?

File: packages/cgb-scripts/template/src/init.php [hunk 5]
```diff
@@ -16,49 +16,50 @@
...
+		'style'         => '<% blockNamePHPLower %>-cgb-style-css', // Enqueue blocks.style.build.css on both frontend & backend.
+		'editor_script' => '<% blockNamePHPLower %>-cgb-block-js', // Enqueue blocks.build.js in the editor only.
+		'editor_style'  => '<% blockNamePHPLower %>-cgb-block-editor-css', // Enqueue blocks.editor.build.css in the editor only.
+	) );
 }
 
-// Hook: Editor assets.
-add_action( 'enqueue_block_editor_assets', '<% blockNamePHPLower %>_cgb_editor_assets' );
+// Hook: Block assets.
+add_action( 'init', '<% blockNamePHPLower %>_cgb_block_assets' );
```
Comment (ahmadawais) [inline packages/cgb-scripts/template/src/init.php:79]: Why did you move to `init`?

Comment (asharirfan) [reply packages/cgb-scripts/template/src/init.php:79]: This PR provides a fix for issue #21 in which it was decided to use `init` hook to enqueue assets. Here are some of the reasons given in the issue:

> Using the `init` hook to register the block type including its assets makes more sense since there is already centralized logic in place that automatically takes care of enqueueing the assets when needed.

> Having this as `init` hook will also allow all the plugin developers to have much better control over which blocks will be registered without passing anything to JS code.

Comment link: https://github.com/ahmadawais/create-guten-block/issues/21#issuecomment-360506088

File: packages/cgb-scripts/template/src/init.php (hunk 4 above)
Comment (asharirfan) [reply packages/cgb-scripts/template/src/init.php]: This looks like a result of the sync that I did last week with the master branch. I will take care of this right away.

Comment (asharirfan) [reply packages/cgb-scripts/template/src/init.php]: Fixed!

**Review by ahmadawais — APPROVED** [review]
Looks good to me! 🔥

@asharirfan go ahead an `Squash and Merge` this PR. Make sure the title is inline with Emojo-log. Then get back to me so I can set you up for releasing this update.

Peace! ✌️

File: packages/cgb-scripts/template/src/init.php [hunk 6]
```diff
@@ -16,49 +16,64 @@
...
+	/**
+	 * Register block styles for both frontend + backend.
+	 */
+	wp_register_style(
 		'<% blockNamePHPLower %>-cgb-style-css', // Handle.
 		plugins_url( 'dist/blocks.style.build.css', dirname( __FILE__ ) ), // Block style CSS.
-		array( 'wp-editor' ) // Dependency to include the CSS after it.
-		// filemtime( plugin_dir_path( __DIR__ ) . 'dist/blocks.style.build.css' ) // Version: File modification time.
+		array( 'wp-editor' ), // Dependency to include the CSS after it.
+		null, // filemtime( plugin_dir_path( __DIR__ ) . 'dist/blocks.style.build.css' ) // Version: File modification time.
```
Comment (ahmadawais) [inline packages/cgb-scripts/template/src/init.php:38]: @asharirfan apparently your code was not tested since it broke in #155 that I just fixed.
